lib/active_merchant/billing/gateways/xpay.rb

```diff
@@ -29,30 +30,35 @@ def initialize(options = {})
       end
 
       def purchase(amount, payment_method, options = {})
-        post = {}
-        add_auth_purchase_params(post, amount, payment_method, options)
         action = options[:operation_id] ? :purchase : :preauth
+        post = add_capture_type(options, action)
+        add_auth_purchase_params(post, amount, payment_method, options)
         commit(action, post, options)
       end
 
       def authorize(amount, payment_method, options = {})
-        post = {}
+        action = options[:operation_id] ? :authorize : :preauth
```

Comment:
@jherreraa parce, could you explain to me this if. Why do you set authorize as action when operation_id is a truethy value? 


The same question applies to all operations.

Reply:
hey! thanks for the question, the logic works this way because of the [2 steps flow](https://developer.nexi.it/en/modalita-di-integrazione/server-to-server/pagamento-2-steps) that implies in the payment step send the operation id.
Let me know if you need any further explanation.

Reply:
we discussed this via huddle and I expected some changes.

Checking the flow preauth is an independent action for the gateway so we need to add it as a function instead of the fuzzy logic of changing the action if the operation_id comes or not.